### DISCUSSION TITLE
Strongly Typed Ids

### DISCUSSION THREAD

**Discussion by DermotMurphy** [riok/mapperly#306] [General]
I am trying to map a Domain Entity (User) to a DTO (UserDto) using this mapper:

[Mapper]
public static partial class UserMapper
{
    [MapProperty(nameof(User.Id.Value), nameof(UserDto.Id))] 
    public static partial UserDto MapUserToDto(User user);
}

public class User
{
    public UserId Id { get; set; }
    public required string FirstName { get; set; }

    public required string LastName { get; set; }

}

public readonly record struct UserId(Guid Value);

public class UserDto
{
    public Guid Id { get; set; }
    public string? FirstName { get; set; }

    public string? LastName { get; set; }

    #region Overrides of Object

    public override string ToString()
    {
        return $"Id: {Id}, FirstName: {FirstName}, LastName: {LastName}";
    }

    #endregion
}

Program.cs looks like this:

var user = new User
{
    Id = new UserId(Guid.NewGuid()),
    FirstName = "James",
    LastName = "Kirk"
};

var dto = UserMapper.MapUserToDto(user);

Console.WriteLine(dto);

I am getting two identical errors:

Severity	Code	Description	Project	File	Line	Suppression State
Error	RMG006	Specified property Value on source type ConsoleMappingObjects.User was not found	ConsoleMappingObjects	E:\Source\Repos\ConsoleMappingObjects\ConsoleMappingObjects\UserMapper.cs	8	Active

Does this MappingAttribute do what I believe it ought to?

[MapProperty(nameof(User.Id.Value), nameof(UserDto.Id))] 

I am expecting the Value property of the Id property on the User object to be mapper to the Id property on the UserDto object.

Am I missing something?

**Comment by latonz** [comment]
The problem is that `nameof` only resolves the symbol it references, for `User.Id.Value` it is `Value` but you want to reference the property path `Id.Value`.

`[MapProperty(nameof(User.Id.Value), nameof(UserDto.Id))]`
resolves to
`[MapProperty("Value", "Id")]`
but what you actually want is
`[MapProperty("Id.Value", "Id")]` or with nameof `[MapProperty(new[] { nameof(User.Id), nameof(User.Id.Value) }, new[] { nameof(UserDto.Id) })]`.

**Comment by DermotMurphy** [comment]
Thanks Lars.

Excellent explanation 👍
